docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+# Peer manager
+
+The peer manager implements the connection policy for the node, based on the
+configuration provided by the operators, the current state of the connections
+reported by the [`Router`](./router.md), and the set of known candidate peers.
+
+This document uses *candidate peer* to refer to the information about a node in
+the network, namely its unique identifier and one or more network addresses.
+This information can be manually configured by the node operator (e.g., via
+`PersistentPeers` parameter) or can be obtained via the Peer-Exchange Protocol
+(PEX), which feeds the peer manager with discovered peers' information.
+
+## Connection policy
+
+The connection policy defines:
+
+1. When the node should establish new connections to peers, and
+1. The next peer to which the router should try to establish a connection.
+
+The first definition is made based the concept of [connection slots](#connection-slots).
+In short, the peer manager will try to fill every connection slot with a peer.
+If all the connection slots are full but there is the possibility to connect to
+a peer that is higher-[ranked](#peer-ranking) that one of the connecting peers,
+the peer manager may attempt to [upgrade a connection slot](#slot-upgrades).
+Details of these operations are provided in the following.
+
+### Connection slots
+
+The number of connection slots is defined by the `MaxConnected` parameter.
+
+While there are available connection slots, the peer manager will provide
+candidate peers to the router, which will try to establish
+new connections to them.
+When the peer manager provides a candidate peer to
+the router, a connection slot becomes _virtually_ occupied by the peer, as the
+router should be dialing it.
+
+When the router establishes a connection to a peer, either
+because it [accepted a connection](#accepted-transition) from a peer,
+or because it [successfully dialed](#dialed-transition) a candidate peer,
+the peer manager should find a slot for this connection.
+
+If there is an available connection slot, and this is the first connection
+established with that peer, the slot is filled by the new connection and
+the peer becomes a [connected peer](#connected-peer).
+The peer manager does not allow two slots to be filled with connections to the
+same peer.
+
+If all `MaxConnected` connection slots are full, the node should _a priori_
+reject the connection with the peer.
```

Comment:
Both. The direction of the connection does not count at this point. Connection established because we dial the peer or because we accepted a connection (at router lever) counts the same.

Actually, the same reasoning also applies to to when we are trying to find a candidate peer to dial.

Reply:
Maybe I missed that in the text. If not, I think it'd be worth clarifying it as often as needed (even if it entails repetition to some degree)

Reply:
I made it clearer and redundant now.